tests/mbt/model/ccv.qnt

```diff
@@ -192,6 +227,87 @@ module ccv_types {
     // given as a pure val so that we can switch cases based on
     // whether a chain is the provider or not
     pure val PROVIDER_CHAIN = "provider"
+
+    // Takes the current provider state and validator set and returns 
+    // the validator set under the current key assignments for the given consumer, as stored in the provider state.
+    pure def applyKeyAssignmentToValSet(
+        providerState: ProviderState,
+        consumer: Chain,
+        valSet: ValidatorSet
+    ): ValidatorSet = {
+        // map each validator in the valSet to a set of
+        // new updates. necessary to allow one validator to generate multiple updates,
+        // which in turn is needed when someone changes keys
+        // (set old key to 0, new key to current power)
+        pure val newValSet = valSet.keys().map(
+            (node) => 
+            pure val power = valSet.get(node)
+            pure val keyAssignmentReplacements = providerState.keyAssignmentReplacements.getOrElse(consumer, Map())
+            if(keyAssignmentReplacements.keys().contains(node)) {
+                // there is a KeyAssignmentReplacement
+                pure val replacement = keyAssignmentReplacements.get(node)
+                pure val oldConsAddr = replacement._1
+                // set old key to 0
+                pure val oldToZero = (oldConsAddr, 0)
+                // set new key to current power
+                pure val newToPower = (providerState.validatorConsumerPubKey.getOrElse(consumer, Map()).getOrElse(node, node), power)
+                Set(oldToZero, newToPower)
+            } else {
+                // no replacement
+                // check if the validator has a key assigned
+                pure val validatorConsumerPubKey = providerState.validatorConsumerPubKey.getOrElse(consumer, Map())
+                if (validatorConsumerPubKey.keys().contains(node)) {
```

Comment:
Why doesn't `keyAssignmentReplacements` have the replacement but `validatorConsumerPubKey` does have a key? What exactly is the relation between `keyAssignmentReplacements` and `validatorConsumerPubKey`?

Reply:
keyAssignmentReplacements has an entry only when the key was replaced in *this block*